app/script/ui/WindowHandler.js

```diff
@@ -0,0 +1,123 @@
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.ui = z.ui || {};
+
+z.ui.WindowHandler = class WindowHandler {
+  constructor() {
+    this.init = this.init.bind(this);
+    this._listen_to_window_resize = this._listen_to_window_resize.bind(this);
+    this.logger = new z.util.Logger('z.ui.WindowHandler', z.config.LOGGER.OPTIONS);
+
+    this.height = 0;
+    this.width = 0;
+
+    this.is_visible = true;
+    this.lost_focus_interval = undefined;
+    this.lost_focus_interval_time = z.config.LOCALYTICS_SESSION_TIMEOUT / 3;
+    this.lost_focus_on = undefined;
+
+    return this;
+  }
+
+  init() {
+    this.width = $(window).width();
+    this.height = $(window).height();
+    this._listen_to_unhandled_promise_rejection();
+    this._listen_to_window_resize();
+    this._listen_to_visibility_change(() => {
+      if (document.visibilityState === 'visible') {
+        this.logger.info('Webapp is visible');
+        this.is_visible = true;
+        window.clearInterval(this.lost_focus_interval);
+        this.lost_focus_interval = undefined;
+        this.lost_focus_on = undefined;
+        amplify.publish(z.event.WebApp.ANALYTICS.START_SESSION);
+      } else {
+        this.logger.info('Webapp is hidden');
+        this.is_visible = false;
+        if (this.lost_focus_interval === undefined) {
+          this.lost_focus_on = Date.now();
+          this.lost_focus_interval = window.setInterval((() => this._check_for_timeout()), this.lost_focus_interval_time);
+        }
+      }
+    });
+    return this;
+  }
+
+  _listen_to_window_resize() {
+    return $(window).on('resize', () => {
+      const current_height = $(window).height();
+      const current_width = $(window).width();
+
+      const change_in_width = this.width - current_width;
+      const change_in_height = this.height - current_height;
+
+      amplify.publish(z.event.WebApp.WINDOW.RESIZE.WIDTH, change_in_width);
+      amplify.publish(z.event.WebApp.WINDOW.RESIZE.HEIGHT, change_in_height);
+
+      this.width = current_width;
+      return this.height = current_height;
+    });
+  }
+
+  _listen_to_unhandled_promise_rejection() {
+    return $(window).on('unhandledrejection', (event) => {
+      const promise_rejection_event = event.originalEvent;
+      const promise_error = promise_rejection_event.reason;
+
+      if (promise_error && promise_error.type === z.conversation.ConversationError.prototype.TYPE.DEGRADED_CONVERSATION_CANCELLATION) {
+        this.logger.log('User has canceled sending a message to a degraded conversation.');
+        promise_rejection_event.preventDefault();
+        promise_rejection_event.stopPropagation();
+        return false;
+      }
+    });
+  }
+
+  _listen_to_visibility_change(callback) {
+    let property_hidden = undefined;
+    let property_visibility_change = undefined;
+
+    if (typeof document.hidden !== 'undefined') {
```

Comment:
This check is no longer neeeded: https://developer.mozilla.org/en-US/docs/Web/API/Document/hidden

All our minimum supported browsers have unprefixed support for document.hidden